src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -144,17 +146,51 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
             }
         }
 
+        public bool TryUpdateOutputWindow(int bytes)
+        {
+            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+        }
+
         public override void Abort(ConnectionAbortedException abortReason)
         {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
             base.Abort(abortReason);
 
             // Unblock the request body.
             RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
         }
 
-        public bool TryUpdateOutputWindow(int bytes)
+        protected override void ApplicationAbort()
         {
-            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+            Log.ApplicationAbortedConnection(ConnectionId, TraceIdentifier);
+            var abortReason = new ConnectionAbortedException(CoreStrings.ConnectionAbortedByApplication);
+            AbortAndReset(abortReason, Http2ErrorCode.CANCEL);
+        }
+
+        public void AbortAndReset(ConnectionAbortedException abortReason, Http2ErrorCode error)
+        {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
+            var outputCompleted = Output.IsCompleted;
```

Comment:
Just put WriteRstStreamAsync on the OutputProducer to avoid any potential race conditions.